docs/proposals/operator-sdk-v1-migration.md

```diff
@@ -0,0 +1,285 @@
+recommend a "lift and shift" approach to migration, versus an in-place update [1]. The scope
+and reach of these changes may be substantial, and it is possible that post migration we will
+not have a clean update path for existing Shipwright users.
+
+[1] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#migration-steps
+
+### Goals
+
+* Identify key decisions that need to be made in the Operator SDK migration
+* Plan the implementation of the Operator SDK migration
+
+### Non-Goals
+
+* Define requirements for a migration tool if we introduce breaking API changes
+* Define a strategy for presenting install/upgrades via OperatorHub
+
+## Proposal
+
+### Project initialization
+
+Operator SDK initializes CRDs with two key flags - `--group` and `--domain`. `domain` refers to
+the DNS segment that all APIs for the operator share, and Operator SDK uses this to boostrap the
+entire project [2]. Our current domain is `build.dev`. We can keep this as is, or change it to 
+`shipwright.io` (or subdomain therein). The latter would be a breaking change, forcing current
+users to migrate their existing Build and BuildStrategy instances.
+
+The project does not require multi-group APIs at present, and moving forward this will likely
+remain the case [3].
+
+[2] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#create-a-new-project
+[3] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#check-if-your-project-is-multi-group
+
+
+### API Migration
+
+Operator SDK v1.x changed the package layout of APIs from 
+`pkg/apis/<group>/<version>/<kind>_types.go` to `/api/version/<kind>_types.go`. Other notable
+changes to API generation include [4]:
+
+1. Replacing `+k8s:deepcopy-gen:interfaces=...` markers with `+kubebuilder:object:root=true`.
+2. Removing `// +k8s:openapi-gen=true` and other related openapi markers. This can be added back,
+   but is not necessary if we do not use `openapi-gen`.
+
+We currently use `deepcopy-gen` markers to generate our deep copy code, and use
+`k8s.io/code-generator` to create our OpenAPI generated code under the covers. Presumably using
+the `kubebuilder` marker will let us retain our generated deep copies.
+
+[4] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#apis
+
+### Controllers
+
+Controller code can more or less be migrated as is from
+`pkg/controller/<kind>/<kind>_controller.go` to `/controllers/<kind>_controller.go` [5]. There are
+a few subtleties with generated types and a change to the manager interface, but in general the
+shift is straightforward.
+
+[5] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#controllers
+
+### RBAC
+
+RBAC is maintained via kubebuilder RBAC Markers, rather than separate YAML files. The RBAC markers
+are placed on the `Reconcile` method of each controller, alongside implementing code. SDK projects
```

Comment:
We probably need to "lift" the content from yaml files and place markers, so that subsequent generation of RBAC yaml in `/config/rbac/` is preserved. ?

Reply:
why to favour markers and not simply rules definition on the RBAC definitions?

Reply:
This is how operator-sdk/kubebuilder maintains RBAC. I get the philosophy here - declare your RBAC within the code that requires the RBAC. Generators then build up the RBAC rules.

Reply:
ok